database/engine2/metric.c

```diff
@@ -0,0 +1,636 @@
+        if(!rc)
+            fatal("DBENGINE METRIC: cannot delete UUID from JudyHS");
+        mrg_stats_size_judyhs_removed_uuid(mrg);
+    }
+
+    // arrayalloc is running lockless here
+    arrayalloc_freez(mrg->index.aral, metric);
+
+    mrg_index_write_unlock(mrg);
+
+    MRG_STATS_DELETED_METRIC(mrg);
+
+    return true;
+}
+
+// ----------------------------------------------------------------------------
+// public API
+
+MRG *mrg_create(void) {
+    MRG *mrg = callocz(1, sizeof(MRG));
+    netdata_rwlock_init(&mrg->index.rwlock);
+    mrg->index.aral = arrayalloc_create(sizeof(METRIC), 65536 / sizeof(METRIC), NULL, NULL, false, true);
+    mrg->stats.size = sizeof(MRG);
+    return mrg;
+}
+
+void mrg_destroy(MRG *mrg __maybe_unused) {
+    // no destruction possible
+    // we can't traverse the metrics list
+
+    // to delete entries, the caller needs to keep pointers to METRICs
+
+    ;
+}
+
+METRIC *mrg_metric_add_and_acquire(MRG *mrg, MRG_ENTRY entry, bool *ret) {
+    // FIXME - support refcounting
+
+//    internal_fatal(entry.latest_time_t > now_realtime_sec(),
+//        "DBENGINE METRIC: metric latest time is in the future");
+
+    return metric_add(mrg, &entry, ret);
+}
+
+METRIC *mrg_metric_get_and_acquire(MRG *mrg, uuid_t *uuid, Word_t section) {
+    // FIXME - support refcounting
+    return metric_get(mrg, uuid, section);
+}
+
+bool mrg_metric_release_and_delete(MRG *mrg, METRIC *metric) {
+    // FIXME - support refcounting
+    return metric_del(mrg, metric);
+}
+
+METRIC *mrg_metric_dup(MRG *mrg __maybe_unused, METRIC *metric) {
+    // FIXME - duplicate refcount
+    return metric;
+}
+
+void mrg_metric_release(MRG *mrg __maybe_unused, METRIC *metric __maybe_unused) {
+    // FIXME - release refcount
+
+}
+
+Word_t mrg_metric_id(MRG *mrg __maybe_unused, METRIC *metric) {
+    return (Word_t)metric;
+}
+
+uuid_t *mrg_metric_uuid(MRG *mrg __maybe_unused, METRIC *metric) {
+    return &metric->uuid;
+}
+
+Word_t mrg_metric_section(MRG *mrg __maybe_unused, METRIC *metric) {
+    return metric->section;
+}
+
+bool mrg_metric_set_first_time_t(MRG *mrg __maybe_unused, METRIC *metric, time_t first_time_t) {
+    netdata_spinlock_lock(&metric->timestamps_lock);
+    metric->first_time_t = first_time_t;
+    netdata_spinlock_unlock(&metric->timestamps_lock);
+
+    return true;
+}
+
+void mrg_metric_expand_retention(MRG *mrg __maybe_unused, METRIC *metric, time_t first_time_t, time_t last_time_t, time_t update_every) {
+    netdata_spinlock_lock(&metric->timestamps_lock);
+
+    if(first_time_t && (!metric->first_time_t || first_time_t < metric->first_time_t))
+        metric->first_time_t = first_time_t;
+
+    if(last_time_t && (!metric->latest_time_t_clean || last_time_t > metric->latest_time_t_clean)) {
+        metric->latest_time_t_clean = last_time_t;
+
+        if(update_every)
+            metric->latest_update_every = update_every;
```

Comment:
```sh
/database/engine2/metric.c: 280 in mrg_metric_expand_retention()
274             metric->first_time_t = first_time_t;
275     
276         if(last_time_t && (!metric->latest_time_t_clean || last_time_t > metric->latest_time_t_clean)) {
277             metric->latest_time_t_clean = last_time_t;
278     
279             if(update_every)
>>>     CID 433484:    (Y2K38_SAFETY)
>>>     A "time_t" value is stored in an integer with too few bits to accommodate it.  The expression "update_every" is cast to "uint32_t".
280                 metric->latest_update_every = update_every;
281         }
282         else if(!metric->latest_update_every && update_every)
283             metric->latest_update_every = update_every;
284     
285         netdata_spinlock_unlock(&metric->timestamps_lock);
```

Reply:
We have the same warning for line `283`.

Reply:
update every again...